Реализован функционал согласно ТЗ спринта 8

pavalka opened the pull request.

pavalka committed:
feat: создать каркас Spring Boot

pavalka committed:
feat: создать класс Film

pavalka committed:
feat: создать класс User

pavalka committed:
feat: создать класс ValidationException

pavalka committed:
feat: создать класс FilmValidator

pavalka committed:
feat: создать класс IdGenerator

pavalka committed:
feat: создать класс FilmController

pavalka committed:
feat: создать класс UserValidator

pavalka committed:
feat: создать класс UserController

pavalka committed:
fix: добавить проверку на null в FilmController.addFilm
fix: добавить проверку на null в FilmController.updateFilm
fix: добавить в FilmController.updateFilm проверку того, что фильм для обновления присутствует в списке фильмов
feat: добавить для FilmController поддержку логирования

pavalka committed:
fix: добавить проверку на null в UserController.addUser
fix: добавить проверку на null в UserController.updateUser
fix: добавить в UserController.updateUser проверку того, что фильм для обновления присутствует в списке фильмов
feat: добавить для UserController поддержку логирования

pavalka committed:
fix: добавить в FilmController.updateFilm вывод в лог сообщения о том, что обновляемый фильм не найден в списке фильмов

pavalka committed:
fix: добавить в FilmValidator.validate проверку того, что releaseDate не превышает текущую дату.

pavalka committed:
test: добавить тесты для класса FilmController

pavalka committed:
feat: создать класс UpdateException

pavalka committed:
fix: исправить в UserValidator.validate ошибку в условии проверки поля login

pavalka committed:
feat: добавить в UserController.updateUser генерацию исключения для случая если id пользователя не найден в списке пользователей.

pavalka committed:
refactor: добавить модификатор final для поля User.name

pavalka committed:
test: создать тесты для класса UserController

pavalka committed:
feat: добавить в FilmController.updateFilm генерацию исключения для случая если id фильма не найден в списке фильмов.

pavalka committed:
feat: добавить в FilmControllerTest тест на обновление фильма с неверным id.

pavalka committed:
refactor: удалить модификаторы static для поля IdGenerator.id и метода IdGenerator.getNextId

pavalka committed:
refactor: добавить аннотацию @ResponseStatus для класса ValidationException

pavalka committed:
refactor: добавить аннотацию @ResponseStatus для класса UpdateException

pavalka committed:
feat: добавить поле UserController.idGenerator
refactor: изменить сигнатуру метода void UserController.addUser(User) на User UserController.addUser(User)
refactor: изменить сигнатуру метода void UserController.updateUser(User) на User UserController.updateUser(User)

pavalka committed:
feat: добавить поле FilmController.idGenerator
refactor: изменить сигнатуру метода void FilmController.addFilm(Film) на Film FilmController.addFilm(Film)
refactor: изменить сигнатуру метода void FilmController.updateFilm(Film) на Film FilmController.updateFilm(Film)

ablohin commented on src/test/java/ru/yandex/practicum/filmorate/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,129 @@
+
+import java.time.LocalDate;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Лучше использовать `import` для каждого класса отдельно, без использования символа астериска. Тем самым ты не столкнешься с проблемами конфликтов слияния при слияние веток при выставлении пул реквеста. В настройках Idea поставить галочку "Use single class import".

ablohin commented on src/test/java/ru/yandex/practicum/filmorate/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,129 @@
+    @Test
+    void addUserShouldThrowExceptionWhenAddUserWithEmptyEmail() {
+        User user = new User("", "user_login", "user_name", LocalDate.of(1990, 3,
+                  12));
```
Перенести кусок выше, там все умещается в 120 символов.

ablohin commented on src/test/java/ru/yandex/practicum/filmorate/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,129 @@
+    @Test
+    void addUserShouldThrowExceptionWhenAddUserWithWrongEmail() {
+        User user = new User("user_email.ru", "user_login", "user_name", LocalDate.of(1990, 3,
+                12));
```
Перенести кусок выше, там все умещается в 120 символов.

ablohin commented on src/test/java/ru/yandex/practicum/filmorate/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,129 @@
+    @Test
+    void addUserShouldThrowExceptionWhenAddUserWithEmptyLogin() {
+        User user = new User("user@email.ru", "", "user_name", LocalDate.of(1990, 3,
+                  12));
```
Перенести кусок выше, там все умещается в 120 символов.

ablohin commented on src/test/java/ru/yandex/practicum/filmorate/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,129 @@
+    @Test
+    void addUserShouldThrowExceptionWhenAddUserWithWrongLogin() {
+        User user = new User("user@email.ru", "user login", "user_name", LocalDate.of(1990,
+                      3, 12));
```
Перенести кусок выше, там все умещается в 120 символов.

pavalka replied:
У меня в IDEA было включено отображение подсказок для параметров метода и с ними не умещалось в 120 символов. Отключил эту опцию...

ablohin commented on src/main/java/ru/yandex/practicum/filmorate/util/IdGenerator.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    public IdGenerator(){
+        this.id = 1;
```
Не нужен здесь `this`.

ablohin commented on src/main/java/ru/yandex/practicum/filmorate/util/IdGenerator.java:
```diff
@@ -0,0 +1,20 @@
+     * @return  значение идентификатора.
+     */
+    public long getNextId() {
+        return id++;
```
При первом вызове будет возращено значение 0. При сохранении в БД с таким идентификатором будет ошибка. Лучше использовать префиксный инкремент. `++i`.

pavalka replied:
При первом вызове метод вернет 1, т. к. при создании объекта id инициализируется 1 .
```
public IdGenerator(){
        id = 1;
    }
```

ablohin commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,62 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.web.bind.annotation.*;
```
Лучше использовать `import` для каждого класса отдельно, без использования символа астериска. Тем самым ты не столкнешься с проблемами конфликтов слияния при слияние веток при выставлении пул реквеста. В настройках Idea поставить галочку "Use single class import". Удалить из таблицы использование пакетов с символом астериска для пакета `util`.

ablohin commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,62 @@
+    private final Map<Long, User> users;
+
+    public UserController() {
+        this.users = new HashMap<>();
```
Не обязательно инициализировать в конструкторе в данном случае, можно сделать сразу в поле класса.

pavalka replied:
Подскажи, пожалуйста, есть какой-то критерий, по которому можно определить где лучше делать инициализацию полей?

ablohin commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,63 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.web.bind.annotation.*;
```
Лучше использовать `import` для каждого класса отдельно, без использования символа астериска. Тем самым ты не столкнешься с проблемами конфликтов слияния при слияние веток при выставлении пул реквеста. В настройках Idea поставить галочку "Use single class import".

ablohin commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,63 @@
+    private final Map<Long, Film> films;
+
+    public FilmController() {
+        this.films = new HashMap<>();
```
Инициализацию можно провести сразу в поле класса `FilmController`.

ablohin commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1 @@
+
```
Раз это файл не нужен, значит его можно удалить. Как появятся проперти, можно будет его добавить.

ablohin commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,134 @@
+
+import java.time.LocalDate;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Лучше использовать `import` для каждого класса отдельно, без использования символа астериска. Тем самым ты не столкнешься с проблемами конфликтов слияния при слияние веток при выставлении пул реквеста. В настройках Idea поставить галочку "Use single class import".

ablohin commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,134 @@
+    void getAllFilmsShouldReturnFilmsCollectionWhenOneFilmIsAdded() {
+        try {
+            filmController.addFilm(new Film("Test film", "Film description", LocalDate.of(2001,
+                    2, 14), 90));
```
Сделай в одну строку, здесь всё влазит в ограничение в 120 символов.

ablohin commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,134 @@
+    @Test
+    void addFilmShouldThrowExceptionWhenAddNull() {
+        assertThrows(ValidationException.class,
+                                        () -> filmController.addFilm(null));
```
Сделай в одну строку, здесь всё влазит в ограничение в 120 символов.

ablohin commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,134 @@
+    @Test
+    void addFilmShouldThrowExceptionWhenAddFilmWithEmptyDescription() {
+        Film film = new Film("Test film", "", LocalDate.of(2001,
+                      2, 14), 90);
```
Сделай в одну строку, здесь всё влазит в ограничение в 120 символов.

ablohin commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,134 @@
+    @Test
+    void addFilmShouldThrowExceptionWhenAddFilmWithDescriptionLengthIs201() {
+        Film film = new Film("Test film", "Description".repeat(18) + "End", LocalDate.of(2001,
+                      2, 14), 90);
```
Сделай в одну строку, здесь всё влазит в ограничение в 120 символов.

ablohin commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,134 @@
+    @Test
+    void addFilmShouldThrowExceptionWhenAddFilmWithWrongDate() {
+        Film film = new Film("Test film", "Description", LocalDate.of(1895, 12,
+                  17), 90);
```
Сделай в одну строку, здесь всё влазит в ограничение в 120 символов.

ablohin commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,134 @@
+    @Test
+    void addFilmShouldThrowExceptionWhenAddFilmWithNegativeDuration() {
+        Film film = new Film("Test Film", "Description", LocalDate.of(2001, 2, 14),
+                             -1);
```
Сделай в одну строку, здесь всё влазит в ограничение в 120 символов.

ablohin commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,134 @@
+    @Test
+    void updateFilmShouldThrowExceptionWhenFilmIdIsWrong() {
+        Film film = new Film("Test film", "Description", LocalDate.of(2001, 2, 14),
+                90);
```
Сделай в одну строку, здесь всё влазит в ограничение в 120 символов.

ablohin commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,134 @@
+        assertDoesNotThrow(() -> filmController.addFilm(film));
+
+        Film newFilm = new Film("Test film", "Description - 1", LocalDate.of(2001, 2,
+                14), 90);
```
Сделай в одну строку, здесь всё влазит в ограничение в 120 символов.

ablohin commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,134 @@
+    @Test
+    void updateFilmShouldUpdateFilm() {
+        Film film = new Film("Test film", "Description", LocalDate.of(2001, 2, 14),
+                             90);
```
Сделай в одну строку, здесь всё влазит в ограничение в 120 символов.

ablohin commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,134 @@
+                             90);
+        assertDoesNotThrow(() -> filmController.addFilm(film));
+        Film newFilm = new Film("Test film", "Description-1", LocalDate.of(2001, 2,
+                     14), 90);
```
Сделай в одну строку, здесь всё влазит в ограничение в 120 символов.

ablohin reviewed (commented):
Молодец, отлично справился. Избавляем от мелких недочетов.

pavalka committed:
refactor: заменить в UserControllerTest.java import static org.junit.jupiter.api.Assertions.* на импорт каждого использованного метода
style: убрать в UserControllerTest перенос строки в строках, где код умещается в 120 сиволов

pavalka committed:
refactor: удалить this в IdGenerator.IdGenerator()

pavalka committed:
refactor: заменить в UserController.java import org.springframework.web.bind.annotation.* на импорт каждой используемой аннотации
refactor: проинициализировать поле UserController.users в момент его объявления
delete: удвлить UserController.UserController()

pavalka committed:
refactor: заменить в FilmController.java import org.springframework.web.bind.annotation.* на импорт каждой используемой аннотации
refactor: проинициализировать поле FilmController.films в момент его объявления
delete: удвлить FilmController.FilmController()

pavalka committed:
delete: удвлить файл application.properties

pavalka committed:
refactor: заменить в FilmControllerTest.java import static org.junit.jupiter.api.Assertions.* на импорт каждого использованного метода
style: убрать в FilmControllerTest перенос строки в строках, где код умещается в 120 сиволов

ablohin commented:
Привет! Всё хорошо, задача принята, дорабатывать не нужно.

ablohin reviewed (approved):
Привет! Всё хорошо, задача принята, дорабатывать не нужно.
